src/core/bitset.c

```diff
@@ -0,0 +1,648 @@
+ *          to query.
+ * \return The allocated capacity.
+ *
+ * \sa \ref igraph_bitset_size().
+ *
+ * Time complexity: O(1).
+ */
+
+
+igraph_integer_t igraph_bitset_capacity(const igraph_bitset_t *bitset) {
+    return IGRAPH_INTEGER_SIZE * (bitset->stor_end - bitset->stor_begin);
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_size
+ * \brief Returns the length of the bitset.
+ *
+ * \experimental
+ *
+ * \param bitset The bitset object
+ * \return The size of the bitset.
+ *
+ * Time complexity: O(1).
+ */
+
+igraph_integer_t igraph_bitset_size(const igraph_bitset_t *bitset) {
+    return bitset->size;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_reserve
+ * \brief Reserves memory for a bitset.
+ *
+ * \experimental
+ *
+ * </para><para>
+ * \a igraph bitsets are flexible, they can grow and
+ * shrink. Growing
+ * however occasionally needs the data in the bitset to be copied.
+ * In order to avoid this, you can call this function to reserve space for
+ * future growth of the bitset.
+ *
+ * </para><para>
+ * Note that this function does \em not change the size of the
+ * bitset. Let us see a small example to clarify things: if you
+ * reserve space for 100 elements and the size of your
+ * bitset was (and still is) 60, then you can surely add additional 40
+ * elements to your bitset before it will be copied.
+ * \param bitset The bitset object.
+ * \param capacity The new \em allocated size of the bitset.
+ * \return Error code:
+ *         \c IGRAPH_ENOMEM if there is not enough memory.
+ *
+ * Time complexity: operating system dependent, should be around
+ * O(n/w),
+ * n is the new allocated size of the bitset,
+ * w is the word size of the machine (32 or 64).
+ */
+
+igraph_error_t igraph_bitset_reserve(igraph_bitset_t *bitset, igraph_integer_t capacity) {
+    igraph_integer_t current_capacity;
+    igraph_integer_t *tmp;
+
+    IGRAPH_ASSERT(bitset != NULL);
+    IGRAPH_ASSERT(bitset->stor_begin != NULL);
+    IGRAPH_ASSERT(capacity >= 0);
+
+    current_capacity = igraph_bitset_capacity(bitset);
+
+    if (IGRAPH_BIT_NSLOTS(capacity) <= IGRAPH_BIT_NSLOTS(current_capacity)) {
+        return IGRAPH_SUCCESS;
+    }
+
+    tmp = IGRAPH_REALLOC(bitset->stor_begin, IGRAPH_BIT_NSLOTS(capacity), igraph_integer_t);
+    IGRAPH_CHECK_OOM(tmp, "Cannot reserve space for bitset.");
+
+    bitset->stor_begin = tmp;
+    bitset->stor_end = bitset->stor_begin + IGRAPH_BIT_NSLOTS(capacity);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_resize
+ * \brief Resizes the bitset.
+ *
+ * \experimental
+ *
+ * </para><para>
+ * Note that this function does not free any memory, just sets the
+ * size of the bitset to the given one. It may, on the other hand,
+ * allocate more memory if the new size is larger than the previous
+ * one. In this case the newly appeared elements in the bitset are
+ * \em not set to zero, they are uninitialized.
```

Comment:
The docs claim that the newly added bits are uninitialized, but the function actually zeros them. While this is different from what `igraph_vector` functions do, I think it is a good thing, as manual zeroing is not trivial. So let's keep the zeroing (agreed @ntamas ?) and change the docs.

Reply:
@Tagl Can you update the doc accordingly please?

Reply:
Updated